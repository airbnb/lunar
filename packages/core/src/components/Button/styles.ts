import { StyleSheet } from '../../hooks/useStyles';

const styleSheet: StyleSheet = ({ color, font, pattern, ui, unit, transition }) => ({
  button: {
    ...pattern.resetButton,
    ...transition.box,
    fontWeight: font.weights.bold,
    position: 'relative',
    color: color.base,
    backgroundColor: color.core.primary[3],
    border: `${ui.borderWidthThick}px solid ${color.core.primary[3]}`,
    borderRadius: ui.borderRadius,
    textAlign: 'center',

    '@selectors': {
      // Removes weird bonus padding from button in Firefox
      '::-moz-focus-inner': {
        border: 0,
        padding: 0,
        margin: 0,
      },

      // Only show hover states on non-disabled
      ':not([disabled]):hover': {
        backgroundColor: color.core.primary[4],
        borderColor: color.core.primary[4],
      },
    },
  },

  button_block: {
    display: 'block',
    width: '100%',
    whiteSpace: 'normal',
    overflow: 'hidden',
  },

  button_borderless: {
    borderColor: 'transparent',

    '@selectors': {
      ':not([disabled]):hover': {
        borderColor: color.accent.bgHover,
      },
    },
  },

  button_disabled: {
    ...pattern.disabled,
  },

  button_invalid: {},

  button_inverted: {
    color: color.core.primary[3],
    backgroundColor: color.accent.bg,

    '@selectors': {
      ':not([disabled]):hover': {
        color: color.core.primary[4],
        backgroundColor: color.accent.bgHover,
      },
    },
  },

  button_loading: {
    cursor: 'default',
  },

  button_small: {
    ...pattern.smallButton,
    minWidth: 6 * unit,
  },

  button_regular: {
    ...pattern.regularButton,
    minWidth: 8 * unit,
  },

  button_large: {
    ...pattern.largeButton,
    minWidth: 9 * unit,
  },
});

export default styleSheet;

export { styleSheet };
