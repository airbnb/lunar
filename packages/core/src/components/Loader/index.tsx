import React from 'react';
import withStyles, { WithStylesProps } from '../../composers/withStyles';

export type Props = {
  /** Display inline instead of absolutely positioned. */
  inline?: boolean;
  /** Invert colors. */
  inverted?: boolean;
  /** Increase the dot size. */
  large?: boolean;
  /** Position statically instead of absolutely. */
  static?: boolean;
};

/** A small 3-dot loading indicator. */
export class Loader extends React.Component<Props & WithStylesProps> {
  static defaultProps = {
    inline: false,
    inverted: false,
    large: false,
    static: false,
  };

  render() {
    const { cx, styles, inline, inverted, large, static: isStatic } = this.props;

    return (
      <div
        className={cx(
          styles.loader,
          inline && styles.loader_inline,
          !isStatic && !inline && styles.loader_absolute,
        )}
      >
        {[1, 2, 3].map(no => (
          <span
            key={no}
            className={cx(
              styles.dot,
              styles[`dot_${no}`],
              large && styles.dot_large,
              inverted && styles.dot_inverted,
            )}
          />
        ))}
      </div>
    );
  }
}

export default withStyles(({ color, unit }) => ({
  loader: {
    margin: '0 auto',
    textAlign: 'center',
  },

  loader_absolute: {
    position: 'absolute',
    top: '50%',
    left: '50%',
    transform: 'translateX(-50%) translateY(-50%)',
  },

  loader_inline: {
    display: 'inline-block',
  },

  dot: {
    width: 6,
    height: 6,
    marginRight: unit / 2,
    borderRadius: '100%',
    display: 'inline-block',
    animationName: {
      name: 'fade',
      '0%, 80%, 100%': {
        opacity: 0,
      },
      '30%, 50%': {
        opacity: 1,
      },
    },
    animationDuration: '0.8s',
    animationIterationCount: 'infinite',
    animationTimingFunction: 'linear',
    animationFillMode: 'both',
    verticalAlign: 'middle',
    backgroundColor: color.core.primary[3],
  },

  dot_inverted: {
    backgroundColor: color.accent.bg,
  },

  dot_large: {
    width: 12,
    height: 12,
    marginRight: unit,
  },

  dot_1: {
    animationDelay: '-0.3s',
  },

  dot_2: {
    animationDelay: '-0.15s',
  },
}))(Loader);
