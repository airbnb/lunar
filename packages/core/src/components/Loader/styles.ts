import { StyleSheet } from '../../hooks/useStyles';

const styleSheet: StyleSheet = ({ color, unit }) => ({
  loader: {
    margin: '0 auto',
    textAlign: 'center',
  },

  loader_absolute: {
    position: 'absolute',
    top: '50%',
    left: '50%',
    transform: 'translateX(-50%) translateY(-50%)',
  },

  loader_inline: {
    display: 'inline-block',
  },

  dot: {
    width: 6,
    height: 6,
    marginRight: unit / 2,
    borderRadius: '100%',
    display: 'inline-block',
    animationName: {
      name: 'fade',
      '0%, 80%, 100%': {
        opacity: 0,
      },
      '30%, 50%': {
        opacity: 1,
      },
    },
    animationDuration: '0.8s',
    animationIterationCount: 'infinite',
    animationTimingFunction: 'linear',
    animationFillMode: 'both',
    verticalAlign: 'middle',
    backgroundColor: color.core.primary[3],
  },

  dot_inverted: {
    backgroundColor: color.accent.bg,
  },

  dot_large: {
    width: 12,
    height: 12,
    marginRight: unit,
  },

  dot_1: {
    animationDelay: '-0.3s',
  },

  dot_2: {
    animationDelay: '-0.15s',
  },
});

export default styleSheet;

export { styleSheet };
