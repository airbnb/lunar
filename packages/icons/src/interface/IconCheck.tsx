import React from 'react';
import withIcon, { Props } from '../withIcon';

function IconCheck(props: Props) {
  return (
    <svg viewBox="0 0 24 24" {...props}>
      <path d="M8.795 15.875l-3.47-3.47a.996.996 0 10-1.41 1.41l4.18 4.18c.39.39 1.02.39 1.41 0l10.58-10.58a.996.996 0 10-1.41-1.41z" />
    </svg>
  );
}

export default withIcon('IconCheck')(IconCheck);
