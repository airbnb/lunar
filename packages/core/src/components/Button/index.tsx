import React from 'react';
import { mutuallyExclusiveTrueProps } from 'airbnb-prop-types';
import withStyles, { WithStylesProps } from '../../composers/withStyles';
import ButtonOrLink, { Props as ButtonOrLinkProps } from '../private/ButtonOrLink';
import Loader from '../Loader';

const sizingProp = mutuallyExclusiveTrueProps('small', 'large');

const GOLDEN_RATIO = (1 + Math.sqrt(5)) / 2;

export type Props = ButtonOrLinkProps & {
  /** Render as a block with full width. */
  block?: boolean;
  /** Render as borderless. */
  borderless?: boolean;
  /** @ignore Hidden prop used in forms. */
  invalid?: boolean;
  /** Invert text colors. */
  inverted?: boolean;
  /** Increase font size to large. */
  large?: boolean;
  /** Decrease font size to small. */
  small?: boolean;
};

/** A standard button and or link for common UI interactions. */
export class Button extends React.Component<Props & WithStylesProps> {
  static propTypes = {
    large: sizingProp,
    small: sizingProp,
  };

  static defaultProps = {
    block: false,
    borderless: false,
    disabled: false,
    invalid: false,
    inverted: false,
    large: false,
    loading: false,
    small: false,
  };

  render() {
    const {
      cx,
      block,
      borderless,
      children,
      disabled,
      invalid,
      inverted,
      large,
      loading,
      small,
      styles,
      ...restProps
    } = this.props;

    return (
      <ButtonOrLink
        {...restProps}
        aria-busy={loading}
        disabled={disabled}
        loading={loading}
        className={cx(
          styles.button,
          large && styles.button_large,
          small && styles.button_small,
          !large && !small && styles.button_regular,
          block && styles.button_block,
          (disabled || loading) && styles.button_disabled,
          (borderless || inverted) && styles.button_inverted,
          invalid && styles.button_invalid,
          borderless && styles.button_borderless,
          loading && styles.button_loading,
        )}
      >
        {loading ? <Loader inline inverted={!inverted} /> : children}
      </ButtonOrLink>
    );
  }
}

export default withStyles(
  ({ color, font, pattern, ui, unit, transition }) => ({
    button: {
      ...pattern.resetButton,
      ...transition.box,
      fontWeight: font.weights.bold,
      position: 'relative',
      color: color.base,
      backgroundColor: color.core.primary[3],
      border: `2px solid ${color.core.primary[3]}`,
      borderRadius: ui.borderRadius,
      textAlign: 'center',

      '@selectors': {
        // Removes weird bonus padding from button in Firefox
        '::-moz-focus-inner': {
          border: 0,
          padding: 0,
          margin: 0,
        },

        // Only show hover states on non-disabled
        ':not([disabled]):hover': {
          backgroundColor: color.core.primary[4],
          borderColor: color.core.primary[4],
        },
      },
    },

    button_block: {
      display: 'block',
      width: '100%',
      whiteSpace: 'normal',
      overflow: 'hidden',
    },

    button_borderless: {
      borderColor: 'transparent',

      '@selectors': {
        ':not([disabled]):hover': {
          borderColor: color.accent.bgHover,
        },
      },
    },

    button_disabled: {
      ...pattern.disabled,
    },

    button_invalid: {},

    button_inverted: {
      color: color.core.primary[3],
      backgroundColor: color.accent.bg,

      '@selectors': {
        ':not([disabled]):hover': {
          color: color.core.primary[4],
          backgroundColor: color.accent.bgHover,
        },
      },
    },

    button_loading: {
      cursor: 'default',
    },

    button_small: {
      ...pattern.smallButton,
      minWidth: GOLDEN_RATIO * 4 * unit,
    },

    button_regular: {
      ...pattern.regularButton,
      minWidth: GOLDEN_RATIO * 6 * unit,
    },

    button_large: {
      ...pattern.largeButton,
      minWidth: GOLDEN_RATIO * 8 * unit,
    },
  }),
  {
    extendable: true,
  },
)(Button);
