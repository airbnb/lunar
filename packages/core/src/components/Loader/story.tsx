import React from 'react';
import Text from '../Text';
import Loader from '.';

export default {
  title: 'Core/Loader',
  parameters: {
    inspectComponents: [Loader],
  },
};

export function aLoaderAbsolutelyCenteredWithinAParent() {
  return (
    <div style={{ height: 100, position: 'relative' }}>
      <Loader />
    </div>
  );
}

aLoaderAbsolutelyCenteredWithinAParent.story = {
  name: 'A loader absolutely centered within a parent.',
};

export function aLoaderInline() {
  return (
    <Text>
      Content before the loader. <Loader inline /> Content after the loader.
    </Text>
  );
}

aLoaderInline.story = {
  name: 'A loader inline.',
};

export function withALargerSizePositionedStatically() {
  return <Loader static large />;
}

withALargerSizePositionedStatically.story = {
  name: 'With a larger size positioned statically.',
};
