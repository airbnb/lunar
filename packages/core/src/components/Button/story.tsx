import React from 'react';
import IconAddAlt from '@airbnb/lunar-icons/lib/interface/IconAddAlt';
import Button from '.';

export default {
  title: 'Core/Button',
  parameters: {
    inspectComponents: [Button],
  },
};

export function standardButtonWithAnInvertedVariant() {
  return (
    <>
      <Button>Button</Button> <Button inverted>Button</Button>
    </>
  );
}

standardButtonWithAnInvertedVariant.story = {
  name: 'Standard button with an inverted variant.',
};

export function renderAnAnchorLinkWhenPassingHref() {
  return (
    <>
      <Button openInNewWindow href="https://github.com/airbnb/lunar">
        Link
      </Button>{' '}
      <Button openInNewWindow inverted href="https://github.com/airbnb/lunar">
        Link
      </Button>
    </>
  );
}

renderAnAnchorLinkWhenPassingHref.story = {
  name: 'Render an anchor link when passing `href`.',
};

export function withEventHandlers() {
  return (
    <Button
      onClick={action('onClick')}
      onMouseOver={action('onMouseOver')}
      onFocus={action('onFocus')}
    >
      Button
    </Button>
  );
}

withEventHandlers.story = {
  name: 'With event handlers.',
};

export function withDifferentSizingSmallRegularDefaultAndLarge() {
  return (
    <>
      <Button small>Button</Button> <Button>Button</Button> <Button large>Button</Button>
      <br />
      <br />
      <Button inverted small>
        Button
      </Button>{' '}
      <Button inverted>Button</Button>{' '}
      <Button inverted large>
        Button
      </Button>
    </>
  );
}

withDifferentSizingSmallRegularDefaultAndLarge.story = {
  name: 'With different sizing: small, regular (default), and large.',
};

export function withDifferentStatesDisabledAndLoading() {
  return (
    <>
      <Button disabled>Button</Button> <Button loading>Button</Button>
      <br />
      <br />
      <Button inverted disabled>
        Button
      </Button>{' '}
      <Button inverted loading>
        Button
      </Button>
    </>
  );
}

withDifferentStatesDisabledAndLoading.story = {
  name: 'With different states: disabled and loading.',
};

export function withBeforeAndOrAfterIcons() {
  return (
    <>
      <div>
        <Button beforeIcon={<IconAddAlt decorative />}>Before icon</Button>{' '}
        <Button afterIcon={<IconAddAlt decorative />}>After icon</Button>
      </div>
    </>
  );
}

withBeforeAndOrAfterIcons.story = {
  name: 'With before and or after icons.',
};

export function withBorderless() {
  return (
    <>
      <div>
        <Button borderless>Button</Button>{' '}
        <Button borderless inverted>
          Button
        </Button>{' '}
        <Button borderless disabled>
          Button
        </Button>
      </div>
    </>
  );
}

withBorderless.story = {
  name: 'With borderless.',
};
