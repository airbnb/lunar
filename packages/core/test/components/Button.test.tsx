import React from 'react';
import { shallowWithStyles } from '@airbnb/lunar-test-utils';
import IconCheck from '@airbnb/lunar-icons/lib/interface/IconCheck';
import Button from '../../src/components/Button';
import Loader from '../../src/components/Loader';

describe('<Button />', () => {
  it('can change `type` of button', () => {
    const wrapper = shallowWithStyles(<Button type="submit">Button</Button>);

    expect(wrapper.prop('type')).toBe('submit');
  });

  it('renders block', () => {
    const wrapper = shallowWithStyles(<Button block>Button</Button>);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders borderless', () => {
    const wrapper = shallowWithStyles(<Button borderless>Button</Button>);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders disabled', () => {
    const wrapper = shallowWithStyles(<Button disabled>Button</Button>);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.prop('disabled')).toBe(true);
  });

  it('renders inverted', () => {
    const wrapper = shallowWithStyles(<Button inverted>Button</Button>);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders large', () => {
    const wrapper = shallowWithStyles(<Button large>Button</Button>);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders loading', () => {
    const child = <div>Button</div>;
    const wrapper = shallowWithStyles(<Button loading>{child}</Button>);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(Loader)).toHaveLength(1);
    expect(wrapper.contains(child)).toBe(false);
  });

  it('renders small', () => {
    const wrapper = shallowWithStyles(<Button small>Button</Button>);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders loading instead of icons', () => {
    const beforeIcon = <IconCheck decorative />;
    const afterIcon = <IconCheck decorative />;
    const wrapper = shallowWithStyles(
      <Button loading beforeIcon={beforeIcon} afterIcon={afterIcon}>
        Default
      </Button>,
    );

    expect(wrapper.contains(beforeIcon)).toBe(false);
    expect(wrapper.contains(afterIcon)).toBe(false);
    expect(wrapper.find(Loader)).toHaveLength(1);
  });

  it('calls wrapped `onMouseUp`', () => {
    const spy = jest.fn();
    const wrapper = shallowWithStyles(<Button onMouseUp={spy}>Button</Button>);

    wrapper.simulate('mouseup', {
      currentTarget: {
        blur() {},
      },
    });

    expect(spy).toHaveBeenCalled();
  });
});
