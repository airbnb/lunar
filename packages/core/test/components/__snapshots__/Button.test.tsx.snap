// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> renders block 1`] = `
<ButtonOrLink
  aria-busy={false}
  className="button_1lp5a5w-o_O-button_regular_un5ukp-o_O-button_block_2ow2so"
>
  Button
</ButtonOrLink>
`;

exports[`<Button /> renders borderless 1`] = `
<ButtonOrLink
  aria-busy={false}
  className="button_1lp5a5w-o_O-button_regular_un5ukp-o_O-button_inverted_1fj8x3s-o_O-button_borderless_1j06z2k"
>
  Button
</ButtonOrLink>
`;

exports[`<Button /> renders disabled 1`] = `
<ButtonOrLink
  aria-busy={false}
  className="button_1lp5a5w-o_O-button_regular_un5ukp-o_O-button_disabled_5k6q5e"
  disabled={true}
>
  Button
</ButtonOrLink>
`;

exports[`<Button /> renders inverted 1`] = `
<ButtonOrLink
  aria-busy={false}
  className="button_1lp5a5w-o_O-button_regular_un5ukp-o_O-button_inverted_1fj8x3s"
>
  Button
</ButtonOrLink>
`;

exports[`<Button /> renders large 1`] = `
<ButtonOrLink
  aria-busy={false}
  className="button_1lp5a5w-o_O-button_large_132nyi"
>
  Button
</ButtonOrLink>
`;

exports[`<Button /> renders loading 1`] = `
<ButtonOrLink
  aria-busy={true}
  className="button_1lp5a5w-o_O-button_regular_un5ukp-o_O-button_disabled_5k6q5e-o_O-button_loading_1ety1lk"
  loading={true}
>
  <Loader
    inline={true}
    inverted={true}
  />
</ButtonOrLink>
`;

exports[`<Button /> renders small 1`] = `
<ButtonOrLink
  aria-busy={false}
  className="button_1lp5a5w-o_O-button_small_ziwd7w"
>
  Button
</ButtonOrLink>
`;
